### PR TITLE
[TimexResolver] Bug fix in numbered week resolution (e.g., 2021-W02)

### PR DESCRIPTION
**Pull request by HiltonGiesenow**
Submitting a fix for https://github.com/microsoft/Recognizers-Text/issues/2428

I added two new tests - one to test for this year, 2021-W02, and one for 2024 where the first day of the year is a Monday (2024-W01). I've confirmed test case DataTypes_Resolver_DateRange_Last_Week is still working (uses "2019-W17").

### CODE REVIEW

File: .NET/Microsoft.Recognizers.Text.DataTypes.DataDrivenTests/TestTimexResolver.cs [hunk 1]
```diff
@@ -525,40 +525,33 @@ public void DataTypes_Resolver_DateTime_TuesAt12PM_UtcInput()
             Assert.AreEqual(nextWeekUtc.ToString("yyyy-MM-dd HH:mm:ss"), resolution.Values[1].Value);
             Assert.IsNull(resolution.Values[1].Start);
             Assert.IsNull(resolution.Values[1].End);
-        }
-
+        }
+
         [TestMethod]
-        public void DataTypes_Resolver_DateTime_Weekend()
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DataTypes.DataDrivenTests/TestTimexResolver.cs:531]: We should keep the weekend and un-anchored month unit tests as well. Can you restore them?
We always suggest appending new tests at the end.

File: .NET/Microsoft.Recognizers.Text.DataTypes.TimexExpression/TimexResolver.cs [hunk 2]
```diff
@@ -236,18 +236,35 @@ private static List<Resolution.Entry> ResolveDuration(TimexProperty timex)
...
+        // this is based on https://stackoverflow.com/questions/19901666/get-date-of-first-and-last-day-of-week-knowing-week-number/34727270
+        private static DateObject FirstDateOfWeek(int year, int weekOfYear, System.Globalization.CultureInfo cultureInfo)
+        {
+            // ISO uses FirstFourDayWeek, and Monday as first day of week, according to https://en.wikipedia.org/wiki/ISO_8601
+            var jan1 = new DateObject(year, 1, 1);
+            int daysOffset = (int)DayOfWeek.Monday - (int)jan1.DayOfWeek;
+            var firstWeekDay = jan1.AddDays(daysOffset);
+
+            int firstWeek = cultureInfo.Calendar.GetWeekOfYear(jan1, System.Globalization.CalendarWeekRule.FirstFourDayWeek, DayOfWeek.Monday);
+            if ((firstWeek <= 1 || firstWeek >= 52) && daysOffset >= -3)
```
Comment (tellarin) [inline .NET/Microsoft.Recognizers.Text.DataTypes.TimexExpression/TimexResolver.cs:260]: Can you also add a unit test to show that W53 works as intended in years that do have it?
Perhaps two test cases, one with reference date at the very end of a year and one with a reference date at the very beginning, but that will still map back to the previous year week 53.

**Review by tellarin — CHANGES_REQUESTED** [review]

File: .NET/Microsoft.Recognizers.Text.DataTypes.TimexExpression/TimexResolver.cs (hunk 2 above)
Comment (HiltonGiesenow) [reply .NET/Microsoft.Recognizers.Text.DataTypes.TimexExpression/TimexResolver.cs:260]: Those missing tests somehow got removed when I updated my local repo - sorry about that - re-added them now. I also put in the new tests you suggested - that what you had in mind?

Comment (tellarin) [reply .NET/Microsoft.Recognizers.Text.DataTypes.TimexExpression/TimexResolver.cs:260]: Seems good. We'll merge once the build passes. Thanks!

Comment (HiltonGiesenow) [reply .NET/Microsoft.Recognizers.Text.DataTypes.TimexExpression/TimexResolver.cs:260]: great stuff, thanks!

**Review by tellarin — APPROVED** [review]

**Pull request merged**